File: src/js/core/chars.js
// 字典（對應 nova_core/chars.py）；資料來自 NOVA_CHARS（data/gen/chars.gen.js）。
(function (Nova) {
  'use strict';
  let cache = null;

  // row: [char, py, stroke, wx, lvl, nameable, meaning, radical]
  function parse() {
    const p = globalThis.NOVA_CHARS;
    const map = new Map();
    for (const r of p.rows) {
      map.set(r[0], {
        char: r[0], py: r[1].split('/'), stroke: r[2], wx: r[3], lvl: r[4],
        nameable: r[5] === 1, meaning: r[6], radical: r[7],
        regular: r[4] <= 2, // 常用字（等級 1–2）
      });
    }
    cache = { map, simp2trad: p.simp2trad, version: p.version, generated: p.generated };
    return cache;
  }
  function data() { return cache || parse(); }
  function allChars() { return data().map; }
  function lookup(ch) { return data().map.get(ch) || null; }

  // 簡體姓氏自動轉繁；回傳 {surname, notes}
  function normalizeSurname(surname) {
    const t = data().simp2trad;
    const notes = [];
    let out = '';
    for (const ch of surname.trim()) {
      if (t[ch]) { out += t[ch]; notes.push('「' + ch + '」已轉為繁體「' + t[ch] + '」'); } else out += ch;
    }
    return { surname: out, notes };
  }

  // 候選字依筆畫分桶；桶內依（等級、碼位）排序，與 Python 一致
  function byStroke(maxLevel, nameableOnly) {
    maxLevel = maxLevel == null ? 2 : maxLevel;
    nameableOnly = nameableOnly == null ? true : nameableOnly;
    const buckets = new Map();
    for (const c of data().map.values()) {
      if (c.lvl > maxLevel || (nameableOnly && !c.nameable)) continue;
      if (!buckets.has(c.stroke)) buckets.set(c.stroke, []);
      buckets.get(c.stroke).push(c);
    }
    for (const list of buckets.values()) {
      list.sort((a, b) => a.lvl - b.lvl || a.char.codePointAt(0) - b.char.codePointAt(0));
    }
    return buckets;
  }

  Nova.chars = { allChars, lookup, normalizeSurname, byStroke, data };
})(globalThis.Nova = globalThis.Nova || {});

File: nova_core/chars.py
"""字典：ETL 產出的 data/gen/chars.json（與瀏覽器端 chars.gen.js 同一份內容）。"""
from __future__ import annotations

import json
from dataclasses import dataclass
from functools import lru_cache

from . import ROOT

GEN_DIR = ROOT / 'data' / 'gen'


@dataclass(frozen=True)
class CharInfo:
    char: str
    py: tuple[str, ...]   # 數字調拼音，第一個為主讀音（臺灣讀音優先）
    stroke: int           # 姓名學筆畫（康熙）
    wx: str               # 五行
    lvl: int              # 1 常見取名用字、2 常用（Big5 常用區）、3 次常用（Big5 次常用區）
    nameable: bool
    meaning: str = ''
    radical: str = ''

    @property
    def regular(self) -> bool:
        """常用字（對應 fate regular）：等級 1–2。"""
        return self.lvl <= 2


@lru_cache(maxsize=None)
def _payload() -> dict:
    with open(GEN_DIR / 'chars.json', encoding='utf-8') as f:
        return json.load(f)


@lru_cache(maxsize=None)
def all_chars() -> dict[str, CharInfo]:
    p = _payload()
    assert p['fields'] == ['char', 'py', 'stroke', 'wx', 'lvl', 'nameable', 'meaning', 'radical']
    out = {}
    for ch, py, stroke, wx, lvl, nameable, meaning, radical in p['rows']:
        out[ch] = CharInfo(ch, tuple(py.split('/')), stroke, wx, lvl, bool(nameable), meaning, radical)
    return out


def simp2trad() -> dict[str, str]:
    return _payload()['simp2trad']


def lookup(ch: str) -> CharInfo | None:
    return all_chars().get(ch)


def normalize_surname(surname: str) -> tuple[str, list[str]]:
    """簡體姓氏自動轉繁；回傳 (繁體姓, 提示訊息)。"""
    table = simp2trad()
    notes = []
    out = []
    for ch in surname.strip():
        if ch in table:
            out.append(table[ch])
            notes.append(f'「{ch}」已轉為繁體「{table[ch]}」')
        else:
            out.append(ch)
    return ''.join(out), notes


def by_stroke(max_level: int = 2, nameable_only: bool = True) -> dict[int, list[CharInfo]]:
    """候選字依筆畫分桶；桶內依（等級、字元）排序以保證確定性。"""
    buckets: dict[int, list[CharInfo]] = {}
    for c in all_chars().values():
        if c.lvl > max_level or (nameable_only and not c.nameable):
            continue
        buckets.setdefault(c.stroke, []).append(c)
    for lst in buckets.values():
        lst.sort(key=lambda c: (c.lvl, c.char))
    return buckets
